api/src/Serializer/PreventAutomaticEmbeddingPropertyMetadataFactory.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace App\Serializer;
+
+use ApiPlatform\Core\Metadata\Property\Factory\PropertyMetadataFactoryInterface;
+use ApiPlatform\Core\Metadata\Property\PropertyMetadata;
+
+/**
+ * This is meant as a thin wrapper around
+ * ApiPlatform\Core\Metadata\Property\Factory\SerializerPropertyMetadataFactory.
+ * That class implements the automatic embedding logic based on serialization groups,
+ * described in https://api-platform.com/docs/core/serialization/#embedding-relations.
+ *
+ * However, we have our own system for serialization groups and embedding entities,
+ * with general groups 'read', 'write', 'create', 'update' etc. instead of entity-
+ * specific groups as they are used in the API platform docs ('book', 'book:update').
+ * Therefore, we don't want relations to be automatically embedded as soon as there
+ * is a property with a matching serialization group on the related entity.
+ *
+ * As an example, we don't want the author to be embedded in the book here:
+ *
+ * #[ApiResource(normalization_context: ['groups' => ['read']])]
+ * class Book {
+ *   #[Groups('read')]
+ *   public ?Person $author = null;
+ * }
+ *
+ * #[ApiResource(normalization_context: ['groups' => ['read']])]
+ * class Person {
+ *   #[Groups('read')]
+ *   public string $name = '';
+ * }
+ *
+ * To prevent the author from being embedded due to just the 'read' group, this
+ * class should be inserted just around SerializerPropertyMetadataFactory in the
+ * decorator chain. It will undo only the undesired changes to the property
+ * metadata that SerializerPropertyMetadataFactory adds.
+ *
+ * Currently, the SerializerPropertyMetadataFactory has a decoration priority of
+ * 30, so this class should be assigned a priority of 29.
+ * https://github.com/api-platform/core/blob/main/src/Bridge/Symfony/Bundle/Resources/config/metadata/metadata.xml#L65
+ */
+final class PreventAutomaticEmbeddingPropertyMetadataFactory implements PropertyMetadataFactoryInterface {
```

Comment:
Bräuchten wir das  nicht, wenn wir vor read etc noch die den Namen der Entity packen würden?

Reply:
Korrekt. Es ist etwas kompliziert, aber der inoffizielle Standard für die Gruppen ist eben, `Period:read` etc. zu verwenden. Dann klappts aber mit dem Embedden nicht mehr ganz so elegant: Um im Camp die Perioden und die Days zu embedden müssten wir dann auf dem Camp den `normalization_context` mit den Gruppen `['Camp:read', 'Period:read', 'Day:read']` einrichten, statt wie vorgeschlagen `['read', 'Camp:Periods', 'Period:Days']`. Die Groups würden also Entities entsprechen statt Relations.

Einen Unterschied macht das, wenn man über verschiedene Relations zur gleichen Entity kommt: Bei ContentNodes könnte man mit den inoffiziell-Standard-Gruppen nur Root, Parent, Children und Descendants gleichzeitig embedden, oder alle nicht embedden. Z.B. nur den Root embedden geht nicht. Da müssten wir dann mit `#[MaxDepth(3)]` oder so arbeiten.

Dafür wäre es beim semi-Standard möglich, im embeddeten Objekt nur einen Teil der Properties auszugeben. Das wäre bei uns aber höchstens in Ausnahmefällen nützlich, etwa wenn wir im Profile Endpoint eine andere Repräsentation des Users embedden wollen als wenn man den User direkt via `/users/1a2b3c4d` abruft.